### ISSUE TITLE
Linking libshark.so on HPC centos 7 with gcc 5.4.0 HDF5 1.8 boost/1.63 gives undefined reference to o `H5::H5Location::openAttribute(std::__cxx11::basic_string ......etc etc

### ISSUE THREAD

**Issue by robertosussex** [ICRAR/shark#1]
Hi there,

I have been trying to install shark on our HPC (Centos 6/7 based)( I am one of the admins). T

The config after

module load   boost/1.63.0  hdf5/serial/gcc/1.8.17 gsl-gcc6.2/2.4  gcc/5.4.0 openmpi/2.0.1/gcc/5.4.0  cmake/3.8.2-gcc-5.4.0

goes well

The compilation goes well however after having created **libshark.so** and during the link of
**shark-importer** gives error:

[ 74%] Building CXX object CMakeFiles/sharklib.dir/src/hdf5/iobase.cpp.o
[ 76%] Building CXX object CMakeFiles/sharklib.dir/src/hdf5/reader.cpp.o
[ 79%] Building CXX object CMakeFiles/sharklib.dir/src/hdf5/traits.cpp.o
[ 81%] Building CXX object CMakeFiles/sharklib.dir/src/hdf5/writer.cpp.o
[ 83%] Linking CXX shared library libshark.so
[ 83%] Built target sharklib
Scanning dependencies of target shark-importer
[ 86%] Building CXX object CMakeFiles/shark-importer.dir/src/importer/descendants.cpp.o
[ 88%] Building CXX object CMakeFiles/shark-importer.dir/src/importer/main.cpp.o
[ 90%] Building CXX object CMakeFiles/shark-importer.dir/src/importer/reader.cpp.o
[ 93%] Building CXX object CMakeFiles/shark-importer.dir/src/importer/velociraptor.cpp.o
[ 95%] Linking CXX executable shark-importer
libshark.so: undefined reference to `H5::H5Location::openAttribute(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
libshark.so: undefined reference to `H5::CommonFG::openGroup(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
libshark.so: undefined reference to `H5::H5File::H5File(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int, H5::FileCreatPropList const&, H5::FileAccPropList const&)'
libshark.so: undefined reference to `H5::CommonFG::createGroup(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long) const'
libshark.so: undefined reference to `_ZNK2H58CommonFG15getObjnameByIdxB5cxx11Ey'
libshark.so: undefined reference to `H5::CommonFG::createDataSet(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, H5::DataType const&, H5::DataSpace const&, H5::DSetCreatPropList const&) const'
libshark.so: undefined reference to `H5::H5Location::setComment(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
libshark.so: undefined reference to `H5::DataSet::write(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, H5::DataType const&, H5::DataSpace const&, H5::DataSpace const&, H5::DSetMemXferPropList const&) const'
libshark.so: undefined reference to `H5::DataType::DataType(H5::PredType const&)'
libshark.so: undefined reference to `H5::Attribute::write(H5::DataType const&, std::__cxx11::basic_string

....

....

===============================

It is obviously something due to libraries consistency/compatibility/location. However, I am not quite clear what. All the link.txt files (that specify the linking options for the different sub directories) contains all the necessary libraries hdf5 libraries -lhdf5_cpp -lhdf5 plus the other ones from boost etc ...

The same version/ src was used in a laptop with a MAC system and compiled and linked fine so the cmake files should be fine.

Any advice ?

Best Regards and Thanks

Dr Roberto Scipioni
MPS Research Computing Support
University of Sussex, UK

**Comment by rtobar** [comment]
Hi Roberto,

tl;dr: it seems you are using different gcc compilers for the hdf5 library and shark.

Thanks for the nice and clear report. As you found out, the necessary libraries are there during link time, but the problem is that they were compiled with a compiler different from the one you are currently using. This is a general interoperability problem between C++ code compiled by different compilers: there are combinations that do work, but some others (even across different versions of the same compiler) that don't.

In particular, your problem seems to suggest you are using gcc > 5.1 (which generates symbols with the `std::__cxx11` prefix for the `string` and `list` classes), but the hdf5 library you are linking against was compiled either with an earlier version of g++, of with a different compiler. On either case the `libhdf5_cpp.so` library actually exists, and the necessary functions are actually there, only that they have different names and cannot be found by your compiler.

What I would suggest you do is that you double-check:
 * Which compiler you are using to compile shark. This should appear as one of the first things you see when running `cmake`. You can also run `make VERBOSE=1` and you will see exactly the compiler you are using. It could happen that even though you loaded the `gcc` module in your HPC environment `cmake` still didn't pick it up, and picked up, say, `/usr/bin/c++` instead or something similar. In that case use the `-DCMAKE_CXX_COMPILER=g++` option to make sure `cmake` picks up your loaded `g++` compiler.
 * How the HDF5 C++ library was compiled. You can run something like this: `objdump -T /path/to/your/libhdf5_cpp.so | c++filt  | grep string` (you can skip the `c++filt` step, it's only for easier reading) to see how the functions are named in the library. If it doesn't have names like `std::__cxx11::basic_string` then it means that it *wasn't* compiled with gcc > 5.1, which explains the problem. In that case you'll need to re-compile the hdf5 library with gcc > 5.1 and try again.

I hope this clarifies things. Let me know if you have any further problems. If all of this guesswork turns to be true then I'll update the docs and add this as a note for future problem-solving.

**Comment by robertosussex** [comment]
Hi there,

Thank you very much for your explanation, I had a feeling it was something of the sort.
I will check things out but from what I know now I think that version of hdf5 was indeed compiled with a version of gcc < 5.4.0 and < 5.1  so that would explain it ...

I will investigate and let you know

Best regards and thanks

Roberto

**Comment by robertosussex** [comment]
Hi there,

ANOTHER ATTEMPT
using modules

                     gcc/6.1.0   
cmake-3.12.1-gcc-6.2.0   
   hdf5-1.10.2-gcc-6.2.0
          gsl-2.5-gcc-6.2.0        
 boost-1.67.0-gcc-6.2.0            

which is a fully consistent stack however the config complains since this version of hdf5 does not have the hdf5_cpp library.. ERROR IS

================

 HDF5 libraries / C++ libraries:  HDF5_hdf5_cpp_LIBRARY-NOTFOUND;/cm/shared/apps/spack/opt/spack/linux-centos7-x86_64/gcc-6.2.0/hdf5-1.10.2-zx5qievuw26cydunbqdk5pptxuultgwn/lib/libhdf5.so / HDF5_hdf5_cpp_LIBRARY-NOTFOUND;/cm/shared/apps/spack/opt/spack/linux-centos7-x86_64/gcc-6.2.0/hdf5-1.10.2-

================

It seems to me to find compatibility is very difficult !!!

Advice ?

I will try another stack

Best regards and thanks


Dr Roberto Scipioni

**Comment by rtobar** [comment]
Hi,

The HDF5 C++ library is indeed a requirement. You previous hdf5 module had it, but the module was compiled with an old compiler; on the other hand the module you are using now is aligned with the compiler you want to use, but doesn't have the C++ library. Is there any chance the module can be re-generated so that it contains that library as well? One would think that there will be other applications that might need the HDF5 C++ library too, so there is potential benefit beyond just having it there just for `shark`.

Worst-case scenario you can also compile hdf5 yourself and have your own hdf5 installation in your home or in a shared directory.

**Comment by robertosussex** [comment]
Hi there

I had to use
module load GSL/2.3-intel-2017a 
Boost/1.63.0-intel-2017a-Python-2.7.13 
HDF5/1.8.18-intel-2017a cmake/3.9.2 
intel/parallel_studio_xe/2017/17.1.132
and the other dependencies they come with within an Easybuild built stack

Now finally it compiled!

Just as an add on would shark benefit from OpenMP ? Any advice on how to include it ?
I did load the module but the config could not see it

Many thanks for your help

Dr Roberto Scipioni
MPS Research Computing Support

**Comment by rtobar** [comment]
Hi!

Glad to hear that you managed to find a combination of modules in your system to compile `shark`. Regarding the OpenMP question, yes, it's beneficial up to some point (around 10 cores), and it then flattens out. In general, but depending on the size of the inputs, the main constrain you'll face is memory consumption, so if you use nodes with many CPUs but limited memory then you have multiple CPUs at the disposal of each `shark` instance.

OpenMP support should be picked up automatically by `cmake` in principle, if your compiler supports it. If you could attach here the output of your `cmake` run then it would be clear what is going on.

**Comment by robertosussex** [comment]
To use the intel stack I had to load intel/parallel_studio_xe/2017/17.1.132 (also for the needed license) which has its own omp5 library which is loaded in the config then giving ...

- Found OpenMP_CXX: -qopenmp  
-- OpenMP found, but is <= 2.0. Compiling without OpenMP support
-- Configuring done

Is there any way to force the use of other libraries? 
Something like

OpenMP_CXX_LIB_NAMES:STRING=  

//Path to a library.
OpenMP_iomp5_LIBRARY:FILEPATH=

in CMakeCache.txt or ?

Best regards and thanks


Dr Roberto Scipioni
MPS Research Computing Support
University of Sussex

**Comment by rtobar** [comment]
That's actually an interesting problem! So far we hadn't experienced it (we don't tend to use Intel compilers much), which is why we don't cater for it. In other words, there's currently not much you can do about it.

After reading a bit online, I think fixing this would require a combination of fixing our `CMakeLists.txt` file, plus require `cmake` >= 3.9. But unluckily enough, I don't have access to a Intel Parallel Studio XE environment to test these changes on. The closest I have is an intel/17.0.4.196 installation (but maybe not as part of the Parallel Studio XE? I'm not sure) which is actually successfully recognizing OpenMP with `cmake` 3.5 and the current state of the code.

If you are willing to undertake the effort, I could write a **tentative** solution (I can't validate it myself!) in a separate branch which you could then try out with a newer `cmake`, and iterate until we get it right. You are currently using `cmake` 3.8, but getting a newer `cmake` shouldn't be an issue (you can even download pre-built binaries from their site). Alternatively if I could get access to your cluster (we could communicate privately for details) I could go over it more quickly and try to provide a proper fix in hopefully less time.

**Comment by robertosussex** [comment]
Thank you 

Let me run some tests then .... I may get back to you 

If you dont hear from me in a few days assume that I moved on
I have created the module shark 
I am sure people wont mind is not supporting openMP

Best regards and thanks


Roberto

**Comment by manodeep** [comment]
@rtobar You should be able to get access to [OzSTAR](https://supercomputing.swin.edu.au/ozstar/). OzSTAR has the Parallel Studio XE (2018).

**Comment by robertosussex** [comment]
Just to let you note thay I used cmake 3.9.2 (see previous mesages) so that part is OK
What do you want me to do then ?
Thanks

Roberto

**Comment by rtobar** [comment]
@manodeep thanks for the suggestion. I actually don't have myself access to OzSTAR, but @cdplagos had, so we used her account and got going. I identified the issue and fixed it, and am about to merge this to the `master` branch.

@robertosussex sorry for the confusion regarding the `cmake` version. I think I got confused with the previous tries, where you were using `cmake` 3.8 at some point. Anyway, please check in a few minutes once I merge this back into the `master`. You should be fine given that you are using `cmake` 3.9; otherwise you'd get a warning explaining the situation and recommending you to upgrade to that version.

**Comment by rtobar** [comment]
Apparently the comment in my commit message automatically marked the issue as closed, so I'm reopening. @robertosussex, please checkout the latest `master` branch and if everything is good please close back the issue.

**Comment by rtobar** [comment]
I will also make a note to improve the validation of the hdf5 library found by cmake, ensure early on when running `cmake` that the hdf5 C++ library is indeed usable, and if it is not then fail with a more useful message explaining the potential problem and its solution.
